src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,150 @@
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private Node<T> current;
+    private int size = 0;
+
     @Override
     public void add(T value) {
+        if (size == 0) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else {
+            Node<T> newNode = new Node<>(tail, value, null);
+            tail.next = newNode;
+            tail = newNode;
+            if (size == 1) {
+                head.next = tail;
+            }
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> currentNode = getNode(index);
+            Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+            if (currentNode.prev == null) {
+                head = newNode;
+            } else {
+                currentNode.prev.next = newNode;
+                currentNode.prev = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        getNode(index);
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index == size || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        getNode(index);
+        T beforeRemove = current.value;
+        current.value = value;
+        return beforeRemove;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        Node<T> remove = getNode(index);
+        if (size == 1) {
+            tail = head = null;
+            size--;
+        } else {
+            removedLogic(remove.value);
+        }
+        return remove.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        return removedLogic(object);
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        int currentIndex = 0;
+        current = head;
+        while (current != null) {
+            if (currentIndex == index) {
+                return current;
+            }
+            current = current.next;
+            currentIndex++;
+        }
+        return null;
+    }
+
+    private boolean removedLogic(T value) {
```

Comment:
let's better pass Node to this method
```suggestion
    private boolean unlink(Node<T> node) {
```